2018/001-external-authentication.md

```diff
@@ -0,0 +1,125 @@
+  systems via Velum.
+* Configure Dex with these connection details, enabling Kubernetes to
+  authenticate, via Dex, against these external services.
+* Support external LDAP, SAML 2.0 and OpenID Connect authentication
+  services.
+
+## Detailed RFC
+
+Kubic/CaaSP currently uses Dex as an authentication intermediary for
+Kubernetes, where Dex is backed by a Kubic managed LDAP server.
+
+Dex supports many authentication "connectors" - LDAP, SAML 2.0, OpenID Connect,
+as well as several less common connectors such as GitHub, LinkedIn, etc. Dex
+additionally supports configuring multiple connectors at once, allowing users
+to choose which service they need to authenticate against.
+
+This RFC proposes allowing Kubic operators to configure arbitary additional
+connectors within Kubic, pointing towards their own authentication backend.
+Initially, this will be limited to Dex's LDAP, SAML 2.0 and OpenID Connect
+connectors.
+
+This RFC does not propose removing that managed LDAP server, though that may
+be something we wish to pursue in the future.
+
+### Proposed change
+
+#### Velum Workflow Changes
+
+We propose to leave the initial bootstrap workflow unchanged, and instead
+allow additional authentication connectors to be configured post-bootstrap, in
+a similar manner to how we allow containter registries to be configured
+post-bootstap.
+
+Operators should be able to login to Velum, navigate to the Settings tab, and
+view, add or remove authentication connectors. Each authentication connector
+type (LDAP, SAML 2.0, OpenID Connect) will have a different view for the view
+and add pages, as Dex requires a different set of configuration inputs for
+each.
+
+It should be possible for operators to configure multiple connectors
+simultaneously, including multiple connectors of the same type (e.g. 2x LDAP).
+
+Once connectors have been configured, Velum will show an "Apply Changes" button,
+again - similar to container registries page, and Salt will be called to push
+out the updated configuration.
+
+#### Salt Implementation
+
+Velum will store the connector details in it's database, and will render the
+values into the Pillar API endpoint for Salt to read.
+
+Once in salt, most of the pieces are already in place for this. We'll need to
+update the Dex ConfigMap template (https://github.com/kubic-project/salt/blob/292b025681915a6ad4d7378624385d231f2f21d4/salt/addons/dex/manifests/15-configmap.yaml#L21)
+to render the new connector details, and then salt will apply the change to
+Kubernetes, causing a roll out of the change.
+
+#### CaaSP CLI Implementation
+
+As we now support SAML 2.0 and OpenID Connect, the current `caasp-cli` workflow
+will no longer be possible. CaaSP CLI must be updated to fully implement the
+OIDC browser based workflow. When a user logs in with `caasp-cli`, they must be
+presented with a URL to open in their browser, where they will login and copy
+the generated token back to a waiting `caasp-cli` in order to complete the
```

Comment:
Will customers still be able to run caasp-cli from a server that has no web browser/gui interface? I know SmartOdds use caasp-cli as part of their CI/CD Jenkins workflow logging in for each build/release and we need to ensure that caasp-cli can still be used without any human interaction n a fully automated way if possible. Having caasp-cli waiting would break these scripts and as such we'd be introducing a breaking change which we must document in velum/somewhere before the changes are made if we can't make caasp-cli fully autonomous. 
Thoughts? I accept I may be misunderstanding things hence want to open a discussion :)

Reply:
The current workflow of entirely automated `caasp-cli login` is problematic as - when you integrate with $random SAML service, you have no ability to predict what data will be needed. For example, maybe the SAML service enforces two factor authentication. This can't be automated away.

The intent is that, caasp-cli will print out the URL the user needs to visit to walk through the authentication flow. This can be visited from any machine - e.g. a server a Texas can print the URL, and I can visit the URL from my desktop in Dublin.

In Kubernetes, the "correct" (in quotes, as there are many correct ways!) is to create a serviceaccount, grant that account the necessary permissions,k and build a kubeconfig which uses these credentials. This flow would give a kubeconfig that is valid for as long as the serviceaccount exists, without the need to touch `caasp-cli` at all.

Reply:
Added a note both about the browser not needing to be the same machine, and about the serviceaccount route for fully automated interactions in the next push.

Reply:
Thanks for explaining. I'm all for the "correct" way with a ServiceAccount and I think we can handle this through documentation and a howto. Once it's setup it can be left and customers CI/CD workflow will remain intact. I think this approach is better than caasp-cli anyway as we've had some reports of instability or random results with caasp-cli.

Service accounts is something customers will be used to for many other integrations we just didn't have the full capability in v1 and early v2 release (which is why some went the caasp-cli route). I'm happy with this approach.